### PR TITLE
Calls info in BaseResponse

### PR DESCRIPTION
**Pull request by zeezdev**
Sometimes it is more convenient to get information about the calls made directly from the mockend request, instead of calculating the order of your calls in the global list (`responses.calls[n]`).

What do you think of such a possibility?

### CODE REVIEW

File: responses/__init__.py
Comment (markstory) [inline responses/__init__.py:1080]: If you have the request and response, why do you need to read out of `self._calls`? Wouldn't `match.calls.add()` work?

Comment (zeezdev) [reply responses/__init__.py:1080]: Actually, that's what I did initially. But then an idea came up:
* reuse an already created call from the "global" list `self._calls`
* have the same call instance in the global list and in the individual response list (`match.calls`)

(using `match.calls.add()` we'll create a separate instance of the same call)

Comment (markstory) [reply responses/__init__.py:1080]: Ok, wanting to reuse the objects seems reasonable to me. What do you think of something like

```python
call = Call(request, response)
self._calls.add_call(call)
match.calls.add_call(call)
```

That would save us having to index/count call objects.

Comment (zeezdev) [reply responses/__init__.py:1078]: Looks good, thanks, Let me try...

Comment (zeezdev) [reply responses/__init__.py]: I replaced the index with 

```python
        call = Call(request, response)
        self._calls.add_call(call)
        match.calls.add_call(call)
```

and added an example of usage Response.calls into the README.

File: README.rst
Comment (beliaev-maksim) [inline README.rst:1099]:
```suggestion
        rsp2 = responses.put(
            "http://foo.bar/42/", json={"id": 42, "name": "Bazz"}
        )
```

File: README.rst
Comment (beliaev-maksim) [inline README.rst:1086]: Need more context on the issue that it solves for the user rather than the statement of technical fact

File: README.rst
Comment (beliaev-maksim) [inline README.rst:1126]: this could be sparse to increase readability

**Review by beliaev-maksim — CHANGES_REQUESTED** [review]
also, CHANGELOG entry is missing

**Comment by zeezdev** [comment]
> also, CHANGELOG entry is missing

Added the CHANGELOG entry.

File: README.rst
Comment (zeezdev) [reply README.rst]: I have extended description and replaced the example, please review.

File: README.rst
Comment (zeezdev) [reply README.rst]: The example has been replaced. Is this still relevant?

**Review by beliaev-maksim — APPROVED** [review]
LGTM

**Comment by markstory** [comment]
This is looking good, a few small formatting issues in the README that need to be fixed up before this can be merged.

**Comment by zeezdev** [comment]
> This is looking good, a few small formatting issues in the README that need to be fixed up before this can be merged.

Fixed.

**Comment by zeezdev** [comment]
It looks like [mypy problems](https://github.com/getsentry/responses/actions/runs/6585106855/job/17907124070?pr=664#step:5:15) is related to v1.6.1 that [released](https://mypy-lang.blogspot.com/2023/10/mypy-16-released.html) several days ago.

I added several ignore comments related to my changes and old code. Please review again.

File: responses/matchers.py
Comment (beliaev-maksim) [inline responses/matchers.py:16]: why is that ?

File: responses/__init__.py
Comment (beliaev-maksim) [inline responses/__init__.py:261]: too many ignores, something looks to be incorrect

Comment (beliaev-maksim) [reply responses/__init__.py:261]: fixed in https://github.com/getsentry/responses/pull/684

first merge https://github.com/getsentry/responses/pull/684, then current PR and we are good to ship the release

Comment (zeezdev) [reply responses/__init__.py]: Merged, thanks!

File: responses/tests/test_responses.py
Comment (beliaev-maksim) [inline responses/tests/test_responses.py:2077]: I think we can make it a bit more readable by enumerating `request` objects as well
```suggestion
        assert rsp.call_count == 2
        
        request1 = rsp.calls[0].request
        assert request.url == "http://www.example.com/"
        assert request.method == "GET"
        
        request2 = rsp.calls[1].request
```

and so on

File: responses/tests/test_responses.py
Comment (beliaev-maksim) [inline responses/tests/test_responses.py:2084]: what exactly do we test here ?

why do we need to test that request body could be converted?

File: responses/tests/test_responses.py
Comment (zeezdev) [reply responses/tests/test_responses.py]: agree, done

File: responses/tests/test_responses.py
Comment (zeezdev) [reply responses/tests/test_responses.py]: (I changed the test a little)

Here we validate that request `rsp2_request1 = rsp2.calls[0].request` obtained from the mock's call contains expected payload `request_payload = {"name": "Bazz"}`, that was used on the put request.

> request body could be converted?

`PreparedRequest` contains only encoded data in the `body` field, so we decode it to compare with the original data - `request_payload`.

Comment (beliaev-maksim) [reply responses/tests/test_responses.py]: > `PreparedRequest` contains only encoded data in the `body` field, so we decode it to compare with the original data - `request_payload`.

correct, but do we need to check it at all as part of `responses` ?

we attach `request` object as is without modification 

https://github.com/getsentry/responses/blob/5c6a3b13bbefc2d4cae0cbd1e9177ab702e15e76/responses/__init__.py#L1041

and in the case of this PR we can skip this check since it does not add a value.

although, you can fire a parallel PR that can add a separate test to validate that we have all the required data is attached to a request
we still will need to discuss if that is required or not but this we can do in a separate thread

File: README.rst
Comment (beliaev-maksim) [inline README.rst:1098]: I understand that this example might look very close to what you have in production. But can we unload it a bit and reduce complexity?

we probably still want to show an example of `concurrent` since this could be one of the main usages, but we do not need all those complicated matrices of settings to show the usage

can you please simplify it to remove additional arguments and have more clear names like
rsp1, rsp2, rsp3

**Review by beliaev-maksim — CHANGES_REQUESTED** [review]
I think we are very close to the conclusion

technically all looks good, just need to simplify user facing docs

File: README.rst
Comment (zeezdev) [reply README.rst:1098]: Got you, I'll try to simplify this.

Comment (zeezdev) [reply README.rst:1098]: simplified, please take a look

File: README.rst
Comment (beliaev-maksim) [inline README.rst:1116]: Can you please update it with f-string

File: README.rst
Comment (beliaev-maksim) [reply README.rst:1098]: @zeezdev now looks a way better, thank you!

**Review by beliaev-maksim — APPROVED** [review]

File: README.rst
Comment (zeezdev) [reply README.rst]: done

**Comment by beliaev-maksim** [comment]
@markstory can you please do a final pass here

File: responses/tests/test_responses.py
Comment (zeezdev) [reply responses/tests/test_responses.py]: @beliaev-maksim and what about this question?

Comment (beliaev-maksim) [reply responses/tests/test_responses.py]: @zeezdev sorry, missed this one in a long thread. Yes.
And if for some reason we have missed the tests that ensure the quality of `Call` objects. Then I would recommend to create separate tests for those and open it in a separate PR.

File: responses/tests/test_responses.py
Comment (beliaev-maksim) [inline responses/tests/test_responses.py:2045]: that comment is probably true only with the default registry, or?

Comment (zeezdev) [reply responses/tests/test_responses.py:2045]: I mean, this mocked resource (`http://www.example.com/2`) isn't requested during the test to explain why we get empty `rsp3.calls`.
I can remove this if it confuses.

Comment (beliaev-maksim) [reply responses/tests/test_responses.py:2045]: that is fine to keep it for the testing purpose.

I just mean that if the user applies a different registry or we switch the default registry (which is unlikely), then this comment will not be correct

https://github.com/getsentry/responses#ordered-registry

Comment (zeezdev) [reply responses/tests/test_responses.py:2045]: but this comment relates to the specific scope of the test. And is it relevant in this case?

If I were to initialize the test with
`@responses.activate(registry=OrderedRegistry)`

then it would be a different case with different expectations, wouldn’t it?

Comment (beliaev-maksim) [reply responses/tests/test_responses.py:2045]: that is more of a note rather than a call for an action :)

I am +1 with the current state of PR

now wait for @markstory to make his round of review

**Comment by beliaev-maksim** [comment]
@zeezdev great work, thanks!

**Pull request merged**
